assets/js/googlesitekit/datastore/site/connection.js

```diff
@@ -0,0 +1,117 @@
+/**
+ * External dependencies
+ */
+import invariant from 'invariant';
+
+/**
+ * Internal dependencies
+ */
+import API from 'googlesitekit-api';
+
+// Actions
+const FETCH_CONNECTION = 'FETCH_CONNECTION';
+const RECEIVE_CONNECTION = 'RECEIVE_CONNECTION';
+const RECEIVE_CONNECTION_FAILED = 'RECEIVE_CONNECTION_FAILED';
+
+export const INITIAL_STATE = {
+	connection: null,
+	isFetchingConnection: false,
+};
+
+export const actions = {
+	fetchConnection() {
+		return {
+			payload: {},
+			type: FETCH_CONNECTION,
+		};
+	},
+
+	receiveConnection( connection ) {
+		invariant( connection, 'connection is required.' );
+
+		return {
+			payload: { connection },
+			type: RECEIVE_CONNECTION,
+		};
+	},
+
+	receiveConnectionFailed( error ) {
+		invariant( error, 'error is required.' );
+
+		return {
+			payload: { error },
+			type: RECEIVE_CONNECTION_FAILED,
+		};
+	},
+};
+
+export const controls = {
+	[ FETCH_CONNECTION ]: () => {
+		return API.get( 'core', 'site', 'connection' );
+	},
+};
+
+export const reducer = ( state, action ) => {
+	switch ( action.type ) {
+		case FETCH_CONNECTION: {
+			return {
+				...state,
+				isFetchingConnection: true,
+			};
+		}
+
+		case RECEIVE_CONNECTION: {
+			const { connection } = action.payload;
+
+			return {
+				...state,
+				isFetchingConnection: false,
+				connection,
+			};
+		}
+
+		case RECEIVE_CONNECTION_FAILED: {
+			const { error } = action.payload;
+
+			return {
+				...state,
+				isFetchingConnection: false,
+				connection: { error, hasError: true },
```

Comment:
This still needs to be changed. A failed request should definitely not affect the actual object that intends to hold the result in a negative way. Depending on what type of error it is (potentially expected negative API response vs there is something truly broken), we may want to put the error into the store object or immediately throw it.

In the cases here I'd lean towards the latter: Getting a connection should never fail as much as resetting should never fail. If it does, this should be thrown so that users would report it to us because it should just never happen.

Reply:
Agreed that we should report this error, but upon reflection: throwing errors from inside the store is unexpected behaviour and frustrating to test. We could easily check for this action being dispatched at the top-level of the app and display an error if we encountered it. At any rate I don't think we should be throwing errors from reducers that propagate up the stack, so for now I've just dispatched the error action and not modified state on error aside from clearing the "isDoing" state.

We shouldn't throw from here, but we should manage errors like this with more thought in the future. Throwing here is difficult to test and atypical, so let's just leave it as a TODO for now.